src/lib/datepicker/datepicker-input.ts

```diff
@@ -244,29 +242,28 @@ export class MatDatepickerInput<D> implements AfterContentInit, ControlValueAcce
       throw createMissingDateImplError('MAT_DATE_FORMATS');
     }
 
+    this._selectionModel = new MatDateSelectionModel<D>(this._dateAdapter);
+    this._selectionModelSubscription = this._selectionModel.onChange
+      .subscribe((selected: MatDateSelectionChange<D>) => {
+        const oldValue = this._prevValue;
+        this.value = <D|null>selected.value;
+        if (!this._dateAdapter.sameDate(oldValue, <D|null>selected.value)) {
```

Comment:
Everything in this if-block should only happen if the change was caused by user interaction. The way it's set up now if the programmer sets the selection model in JS it will trigger this

Reply:
I'm starting to think, that I should call `selectionModel.select()` only in `DatepickerInput`. I should also return back `MatDatepicker.onChange()` and subscribe to it. MatCalendar and others should call `selectionModel.select()`. Instead, they should pass events as it was before selectionModel. If it is right. What is the purpose of selection model? It will not do anything at all.

Reply:
Let me give 2 examples of how I imagine it working:

1. Single date selection: user clicks a date and the calendar calls `selectionModel.select(date)` with the date they picked. The selection model knows it's in single selection mode and so it indicates `selectionFinished: true` on the event it emits. The calendar sees that selection is finished and closes itself.

2. Range selection: user clicks a date and the calendar calls `selectionModel.select(date)` with the date they picked. The selection model knows it's in range selection mode so it indicates `selectionFinished: false` on the event it emits. The calendar sees that selection is not finished so it does not close. The user clicks another date and the calendar again calls `selectionModel.select(date)`. The selection model sees that the user has now picked both a start and end date so it sets `selectionFinished: true` on the event it emits. The calendar sees this and closes itself.

Does this help? I want the calendar to not have to know what to do when the user clicks, it just passes it to the selection model and lets the selection model decide

Reply:
This doesn't help. I've done `selectionModel.select(date)`, but as you said, it causes emits in if-block, and it is wrong. 
I can see only one solution: add new boolean ReplaySubject. if it's value true - it means I should emit event. if not, I shouldn't. How should I guess, new selected value set by calendar or the programmer set it up?

Reply:
Maybe subscribing to the `Observable` isn't the best way to do this, why don't you just have the `select` method return a `SelectionResult` object (that would hold the same info your event is currently holding)